internal/vulnstore/postgres/gc.go

```diff
@@ -0,0 +1,265 @@
+	sem := semaphore.NewWeighted(int64(runtime.NumCPU()))
+
+	errC := make(chan error, len(updaters))
+
+	for _, updater := range updaters {
+		err = sem.Acquire(ctx, 1)
+		if err != nil {
+			break
+		}
+		go func(u string) {
+			defer sem.Release(1)
+			err := chunkedCleanup(ctx, s.pool, u)
+			if err != nil {
+				errC <- err
+			}
+		}(updater)
+	}
+
+	// unconditionally wait for all in-flight go routines to return.
+	// the use of context.Background and lack of error checking is intentional.
+	// all in-flight go routines are guarantee to release their sems.
+	sem.Acquire(context.Background(), int64(runtime.NumCPU()))
+
+	close(errC)
+	for e := range errC {
+		if e != nil {
+			return totalOps - deletedOps, fmt.Errorf("received error issuing deletes: %v", err)
+		}
+	}
+	return totalOps - deletedOps, nil
+}
+
+// distinctUpdaters returns all updaters which have registered an update
+// operation.
+func distinctUpdaters(ctx context.Context, pool *pgxpool.Pool) ([]string, error) {
+	const (
+		// will always contain at least two update operations
+		selectUpdaters = `
+SELECT DISTINCT(updater) FROM update_operation;
+`
+	)
+	rows, err := pool.Query(ctx, selectUpdaters)
+	defer rows.Close()
+	if err != nil {
+		return nil, fmt.Errorf("error selecting distinct updaters: %v", err)
+	}
+
+	var updaters []string
+	for rows.Next() {
+		var updater string
+		err := rows.Scan(&updater)
+		switch err {
+		case nil:
+			// hop out
+		default:
+			return nil, fmt.Errorf("error scanning updater: %v", err)
+		}
+		updaters = append(updaters, updater)
+	}
+	return updaters, nil
+}
+
+// eligibleUpdateOpts returns a list of update operation refs which exceed the specified
+// keep value.
+func eligibleUpdateOpts(ctx context.Context, pool *pgxpool.Pool, keep int) ([]uuid.UUID, int64, error) {
+	const (
+		// this query will return rows of UUID arrays.
+		// each returned array are the UUIDs which exceed the provided keep value
+		updateOps = `
+WITH ordered_ops AS (
+    SELECT array_agg(ref ORDER BY date DESC) AS refs FROM update_operation GROUP BY updater
+)
+SELECT ordered_ops.refs[$1:]
+FROM ordered_ops
+WHERE array_length(ordered_ops.refs, 1) > $2;
+`
+	)
+
+	var total int64
+
+	// gather any update operations exceeding our keep value.
+	// keep+1 is used because PG's array slicing is inclusive,
+	// we want to grab all items once after our keep value.
+	m := []uuid.UUID{}
+	rows, err := pool.Query(ctx, updateOps, keep+1, keep)
+	switch err {
+	case nil:
+	default:
+		return nil, total, fmt.Errorf("error querying for update operations: %v", err)
+	}
+	defer rows.Close()
+	for rows.Next() {
+		// pgx will not scan directly into a []uuid.UUID
+		tmp := pgtype.UUIDArray{}
+		err := rows.Scan(&tmp)
+		if err != nil {
+			return nil, total, fmt.Errorf("error scanning update operations: %w", err)
+		}
+		for _, u := range tmp.Elements {
+			total++
+			m = append(m, u.Bytes) // this works since [16]byte value is assignable to uuid.UUID
+		}
+	}
```

Comment:
need to check `rows.Err()`

Reply:
Still a little confused why that's necessary.

Isn't this only useful if we *don't* defer the rows.Close() ? rows.err is set on Close() https://github.com/jackc/pgx/blob/master/rows.go#L132 

In other cases rows.err is returned to you directly: https://github.com/jackc/pgx/blob/master/conn.go#L608

https://github.com/jackc/pgx/blob/master/rows.go#L76

Reply:
explained out of band. the checks are needed.